### PR TITLE
AIP-84 Get Task Instance Tries

### PR DESCRIPTION
**Pull request by kandharvishnu**
related to: https://github.com/apache/airflow/issues/42370

---

### CODE REVIEW

File: airflow/api_fastapi/core_api/routes/public/task_instances.py [hunk 1]
```diff
@@ -234,6 +236,69 @@ def get_task_instance_dependencies(
...
+            f"The Task Instance with dag_id: `{dag_id}`, run_id: `{dag_run_id}`, task_id: `{task_id}` and map_index: `{map_index}` was not found",
+        )
+    task_instances = [
+        TaskInstanceHistoryResponse.model_validate(task_instance, from_attributes=True)
+        for task_instance in task_instance_select
+    ]
+    return TaskInstanceHistoryCollectionResponse(
+        task_instances=task_instances,
+        total_entries=len(task_instances),
+    )
```
Comment (rawwar) [inline airflow/api_fastapi/core_api/routes/public/task_instances.py]:
```suggestion
    return TaskInstanceHistoryCollectionResponse(
        task_instances=task_instance_select,
        total_entries=len(task_instance_select),
    )
```

File: tests/api_fastapi/core_api/routes/public/test_task_instances.py [hunk 2]
```diff
@@ -2318,3 +2318,44 @@ def test_raises_404_for_non_existent_dag(self, test_client):
...
+        self.create_task_instances(
+            session=session, task_instances=[{"state": State.SUCCESS}], with_ti_history=True
+        )
+        print("here")
+        response = test_client.get(
+            "/public/dags/example_python_operator/dagRuns/TEST_DAG_RUN_ID/taskInstances/print_the_context/tries"
+        )
+        assert response.status_code == 200
+        assert response.json()["total_entries"] == 2  # The task instance and its history
+        assert len(response.json()["task_instances"]) == 2
```
Comment (rawwar) [inline tests/api_fastapi/core_api/routes/public/test_task_instances.py:2334]: can we also check the task_instance details?

File: airflow/api_fastapi/core_api/routes/public/task_instances.py [hunk 3]
```diff
@@ -234,6 +236,69 @@ def get_task_instance_dependencies(
...
+            orm_object.task_id == task_id,
+            orm_object.map_index == map_index,
+        )
+        return query
+
+    # Exclude TaskInstance with state UP_FOR_RETRY since they have been recorded in TaskInstanceHistory
+    tis = session.scalars(
+        _query(TI).where(or_(TI.state != TaskInstanceState.UP_FOR_RETRY, TI.state.is_(None)))
+    ).all()
+    task_instance_select = session.scalars(_query(TIH)).all() + tis
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/core_api/routes/public/task_instances.py]: The request has been made, it's not a select anymore, but `task_instances`

File: tests/api_fastapi/core_api/routes/public/test_task_instances.py [hunk 4]
```diff
@@ -2318,3 +2318,44 @@ def test_raises_404_for_non_existent_dag(self, test_client):
...
+    def test_raises_404_for_nonexistent_task_instance(self, test_client, session):
+        self.create_task_instances(session)
+        response = test_client.get(
+            "/public/dags/example_python_operator/dagRuns/TEST_DAG_RUN_ID/taskInstances/non_existent_task/tries"
+        )
+        assert response.status_code == 404
+
+        assert response.json() == {
+            "detail": "The Task Instance with dag_id: `example_python_operator`, run_id: `TEST_DAG_RUN_ID`, task_id: `non_existent_task` and map_index: `-1` was not found"
+        }
```
Comment (pierrejeambrun) [inline tests/api_fastapi/core_api/routes/public/test_task_instances.py:2443]: I think we are missing a test for `mapped_task_instances`. It was named `test_mapped_task_should_respond_200`

**Review by pierrejeambrun — COMMENTED** [review]
Nice.

Just a few suggestions.

File: tests/api_fastapi/core_api/routes/public/test_task_instances.py (hunk 4 above)
Comment (pierrejeambrun) [reply tests/api_fastapi/core_api/routes/public/test_task_instances.py:2443]: No we are not. mapped is in the following PR. resolving.

Comment (kandharvishnu) [reply tests/api_fastapi/core_api/routes/public/test_task_instances.py:2443]: It is in another PR - https://github.com/apache/airflow/pull/44303

**Review by pierrejeambrun — APPROVED** [review]
Ready to merge, thanks.

**Comment by pierrejeambrun** [comment]
Rebasing again, need latest main fix for the CI.

**Pull request merged**